File: scripts/ui/fragments/CataclysmFragment.js
/** a simple collapser for cataclysms*/
module.exports = extend(Fragment, {
    consequences: [],
    title: "",
    text: "",
    duration: 0,

    collapse(title, duration, consequences){
        this.title = title;
        this.duration = duration;
        this.consequences = consequences;

        this.buildText();
    },
    buildText(){
        this.text = Core.bundle.get("stat.consequence") + ": ";

        if(this.consequences.length > 0){
            this.consequences.forEach(e => {
                this.text += e.name + (this.consequences.indexOf(e) != this.consequences.length - 1 ? ", " : "");
            });
        } else this.text += "[lightgray]" + Core.bundle.get("none");
    },
    build(parent){
        parent.fill(null, t => {
            t.top();
            t.visible = Vars.ui.hudfrag.shown;
    
            t.collapser(v => v.add().height(38), () => Vars.state.isPaused() || Core.settings.getBool("displayCoreItems")).row();
    
            t.table(null, b => {
                b.collapser(top => top.table(null, m => {

                    m.table(null, tt => {
                        tt.background(Styles.black6).add("").pad(8).update(label => {

                            label.color.set(Color.orange).lerp(Color.scarlet, Mathf.absin(Time.time, 2, 1))
                            label.setText(this.title);
        
                        })
                    });

                    m.row();

                    m.table(null, tt => {
                        tt.add("").pad(1).update(label => {
                            label.setText(this.text);
                            label.setFontScale(0.83);
                        });
                    });

                }), true, () => {
                    if(!Vars.ui.hudfrag.shown || Vars.state.isPaused()) return false;
                    if(Vars.state.isMenu()){
                        this.duration = 0
                        return false;
                    }
                    return (this.duration -= Time.delta) > 0;
                })
                .touchable(Touchable.disabled)
                .fillX().row();
            });
        });
    }
});

File: src/animalia/ui/fragments/CataclysmFragment.java
package animalia.ui.fragments;

import arc.Core;
import arc.graphics.Color;
import arc.math.Mathf;
import arc.scene.Group;
import arc.scene.event.Touchable;
import arc.util.Time;
import mindustry.Vars;
import mindustry.ui.Styles;
import animalia.type.Consequence;

public class CataclysmFragment {
    public Consequence[] consequences = {};
    public String title = "", text = "";
    float duration = 0f;

    public void build(Group parent){
        parent.fill(t -> {
            t.top();
            t.visible = Vars.ui.hudfrag.shown;

            t.collapser(v -> v.add().height(38), () -> Vars.state.isPaused() || Core.settings.getBool("displayCoreItems")).row();

            t.table(b -> {
                b.collapser(top -> top.table(m -> {
                    m.table(tt -> {
                        tt.background(Styles.black6).add("").pad(8).update(label -> {
                            label.color.set(Color.orange).lerp(Color.scarlet, Mathf.absin(Time.time, 2, 1));
                            label.setText(title);
                        });
                    });
                m.row();

                m.table(tt -> {
                    tt.add("").pad(1).update(label -> {
                        label.setText(text);
                        label.setFontScale(0.83f);
                    });
                });

                }), true, () -> {
                    if(!Vars.ui.hudfrag.shown || Vars.state.isPaused()) return false;
                    if(Vars.state.isMenu()){
                        duration = 0;
                        return false;
                    }
                    return (duration -= Time.delta) > 0;
                })
                .touchable(Touchable.disabled)
                .fillX().row();
            });
        });
    }

    public void build(){
        text = Core.bundle.get("stat.consequence") + ": ";

        if(consequences.length > 0){
            for(Consequence c : consequences) text += c.name + ", ";
            text = text.substring(0, text.length() - 2);
        } else text += "[lightgray]" + Core.bundle.get("none");
    }
    public void collapse(String title, float duration, Consequence[] consequences){
        this.title = title;
        this.duration = duration;
        this.consequences = consequences;

        build();
    }
}
